### PR TITLE
MDEV-37939 Add function context to division by zero warnings

### PR DESCRIPTION
**Pull request by alexaustin007**
JIRA: https://jira.mariadb.org/browse/MDEV-37939

  Summary
  - Add contextual “Division by 0 in function” warnings for LOG/LN/LOG2/LOG10
  - Keep warning code ER_DIVISION_BY_ZERO (1365) unchanged
  - Add a regression test for LOG-family warnings vs generic division warnings

  Testing
  - MTR_BINDIR=../build ./mariadb-test-run.pl --suite=main --record func_math_div_zero
  - MTR_BINDIR=../build ./mariadb-test-run.pl --suite=main func_math_div_zero

### CODE REVIEW

File: sql/item_func.cc [hunk 1]
```diff
@@ -793,6 +793,20 @@ void Item_func::signal_divide_by_null()
 }
 
 
+void Item_func::signal_divide_by_null(const char *context)
+{
```
Comment (gkodinov) [inline sql/item_func.cc:800]: Why keep the old one (no arguments)? 
I'd say remove it: it's not being used anymore.

File: sql/item_func.h [hunk 2]
```diff
@@ -251,6 +251,7 @@ class Item_func :public Item_func_or_sum
   String *val_str_from_val_str_ascii(String *str, String *str2);
 
   void signal_divide_by_null();
```
Comment (gkodinov) [inline sql/item_func.h:253]: ditto here: remove the old one.
Also, please consider moving this under "protected": let's make sure it's not used anywhere externally. 
I'd even go as far as making it an inline static function in item_func.cc: it doesn't have to be a class method. I'd at least make it static to highlight that it's just a utility function.

File: mysql-test/main/func_math_div_zero.result [hunk 3]
```diff
@@ -0,0 +1,135 @@
+#
+# MDEV-37939: Include function name in division by zero warnings
+#
+#
+# Test with ERROR_FOR_DIVISION_BY_ZERO enabled
```
Comment (gkodinov) [inline mysql-test/main/func_math_div_zero.result]: Please add coverage for all of the functions you're affecting, namely:

* Item_func_div (decimal, integer and real arguments with numerical and string arguments)
* Item_func_mod (int, decimal and real results)
* Item_func_ln (you have that)
* Item_func_log (you have that)
* Item_func_log2 
* Item_func_log10

File: sql/share/errmsg-utf8.txt [hunk 4]
```diff
@@ -12362,3 +12362,5 @@ ER_WARN_CONFLICTING_COMPOUND_INDEX_HINT_FOR_KEY
         eng "Hint %s is ignored as conflicting/duplicated (an index hint of the same type or opposite kind has already been specified for the key)"
 ER_WARN_NO_IMPLICIT_QB_NAMES_IN_VIEW
 	eng "Implicit query block names are ignored for hints specified within VIEWs"
+ER_DIVISION_BY_ZERO_IN_FUNC 22012
+        eng "Division by 0 in function %s"
```
Comment (gkodinov) [inline sql/share/errmsg-utf8.txt]: Since you are at it, I'd even add a specific cause for the error. Some functions could throw this error on several occasions.

File: mysql-test/main/func_math_div_zero.result [hunk 5]
```diff
@@ -0,0 +1,135 @@
+#
```
Comment (gkodinov) [inline mysql-test/main/func_math_div_zero.result:1]: Please observe https://mariadb.org/get-involved/getting-started-for-developers/submitting-pull-request/#make-sure-your-commit-messages-and-content-follow-the-following-guidelines when compiling the commit message text.

**Review by gkodinov — CHANGES_REQUESTED** [review]
This is a preliminary review. The approach is sound IMHO, but there's some things missing (see the comments below).

Also, please make sure you re-record the rest of the failing test cases too.

**Comment by alexaustin007** [comment]
Thank you for the suggestion @gkodinov . I removed the no arg signal_divide_by_null() and made the remaining method protected
All call sites now pass explicit context and also expanded func_math_div_zero` to cover DIV/MOD with decimal/string args and re‑recorded the result file. 
I still  haven't added the specific cause for the error, once this looks good, I can cover up in my next steps.

File: sql/share/errmsg-utf8.txt (hunk 4 above)
Comment (alexaustin007) [reply sql/share/errmsg-utf8.txt]: Just to let you know, I haven’t touched this in the current commit

Comment (gkodinov) [reply sql/share/errmsg-utf8.txt]: To get the final review approval, I think you also should address Sergei's question.

**Comment by alexaustin007** [comment]
Updated it so LOG/LN/LOG2/LOG10 now emit the correct warning (Invalid argument for logarithm) instead of Division by 0. Division/DIV/MOD remain unchanged also reordered the expanded test of func_math_div_zero to verify empty warnings and all cases. Build and test pass. @gkodinov @vuvova Requesting to review. Thank you

File: sql/share/errmsg-utf8.txt (hunk 4 above)
Comment (alexaustin007) [reply sql/share/errmsg-utf8.txt]: The current changes should address Sergei's question now.

File: sql/item_func.cc [hunk 6]
```diff
@@ -783,12 +783,36 @@ void Item_udf_func::fix_num_length_and_dec()
 #endif
 
 
+namespace {
+
+void push_divide_by_zero_warning(THD *thd)
```
Comment (gkodinov) [inline sql/item_func.cc]: It's a but of an overkill these helper functions. push_warning itself is not considered too much IMHO. It's OK to leave it like this. Just saying I wouldn't have went to the trouble. I was hoping that you can either abandon Item_func::signal_... altogether or use local static helper functions. Now you're doing both. This is definitely redundant.

File: sql/item_func.cc [hunk 7]
```diff
@@ -783,12 +783,36 @@ void Item_udf_func::fix_num_length_and_dec()
...
+{
+  push_warning(thd, Sql_condition::WARN_LEVEL_WARN, ER_DIVISION_BY_ZERO,
+               ER_THD(thd, ER_DIVISION_BY_ZERO));
+}
+
+void push_invalid_log_argument_warning(THD *thd)
+{
+  push_warning(thd, Sql_condition::WARN_LEVEL_WARN,
+               ER_INVALID_ARGUMENT_FOR_LOGARITHM,
+               ER_THD(thd, ER_INVALID_ARGUMENT_FOR_LOGARITHM));
```
Comment (gkodinov) [inline sql/item_func.cc]: Heh, you've found yourself a dead error code! This code is not used anywhere currently. 
I'd personally prefer to leave it like that and add a new one that actually helps, e.g. "passing a negative argument to ln produces undefined results".

Otherwise we're back to the situation before your change: you get a cryptic error message that you can't really pinpoint to any specific part of a complex expression.  

I will try to offer what I think is a meaningful error message in each case below.

File: sql/item_func.cc [hunk 8]
```diff
@@ -2021,7 +2045,7 @@ double Item_func_ln::val_real()
     return 0.0;
   if (value <= 0.0)
   {
-    signal_divide_by_null();
+    signal_invalid_log_argument();
```
Comment (gkodinov) [inline sql/item_func.cc]: Message: Passing negative argument to ln() makes the function's result undefined.

File: sql/item_func.cc [hunk 9]
```diff
@@ -2041,7 +2065,7 @@ double Item_func_log::val_real()
...
```
Comment (gkodinov) [inline sql/item_func.cc]: Message: Passing negative argument to log() makes the function's result undefined.

File: sql/item_func.cc [hunk 10]
```diff
@@ -2051,7 +2075,7 @@ double Item_func_log::val_real()
       return 0.0;
     if (value2 <= 0.0 || value == 1.0)
     {
-      signal_divide_by_null();
+      signal_invalid_log_argument();
```
Comment (gkodinov) [inline sql/item_func.cc]: Message: Passing negative 2nd argument to log() makes the function's result undefined.

File: sql/item_func.cc [hunk 11]
```diff
@@ -2068,7 +2092,7 @@ double Item_func_log2::val_real()
     return 0.0;
   if (value <= 0.0)
   {
-    signal_divide_by_null();
+    signal_invalid_log_argument();
```
Comment (gkodinov) [inline sql/item_func.cc]: Message: Passing negative argument to log2() makes the function's result undefined.

File: sql/item_func.cc [hunk 12]
```diff
@@ -2082,7 +2106,7 @@ double Item_func_log10::val_real()
...
```
Comment (gkodinov) [inline sql/item_func.cc]: Message: Passing negative argument to log10() makes the function's result undefined.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Something interesting from https://dev.mysql.com/doc/refman/8.4/en/mathematical-functions.html#function_log2:

> If X is less than or equal to 0.0E0, the function returns NULL and a warning “Invalid argument for logarithm” is reported. Returns NULL if X is NULL.

I think this is a good idea. But I will leave this to the final reviewer.

**Comment by alexaustin007** [comment]
@gkodinov Updated the log warnings to use specific messages instead of "division by 0", removed the dead code, and refreshed the tests/results. Requesting to review

File: sql/item_func.cc (hunk 8 above)
Comment (alexaustin007) [reply sql/item_func.cc]: added non-positive instead of negative since value <= 0, Let me know if it needs any change @gkodinov

File: sql/item_func.h [hunk 13]
```diff
@@ -249,8 +249,10 @@ class Item_func :public Item_func_or_sum
     return null_value;
   }
   String *val_str_from_val_str_ascii(String *str, String *str2);
-
+protected:
   void signal_divide_by_null();
+  void signal_log_domain_error(uint sql_errno);
```
Comment (gkodinov) [inline sql/item_func.h]: make this a static func please. No need for it to be an instance method.

**Review by gkodinov — APPROVED** [review]
LGTM. Please wait for the final review.

File: sql/item_func.h (hunk 13 above)
Comment (alexaustin007) [reply sql/item_func.h]: Its a Static function now. Thank you. @gkodinov requesting to review

Comment (gkodinov) [reply sql/item_func.h]: LGTM

**Review by gkodinov — CHANGES_REQUESTED** [review]
There are failing tests because of the new warnings. Please fix.

**Comment by alexaustin007** [comment]
> There are failing tests because of the new warnings. Please fix.

Re-recorded test results for the updated LOG domain warnings. All related tests now pass. The main.tmp_space_usage failure in the MSAN build seems unrelated issue

**Comment by gkodinov** [comment]
Thanks, there's also some conflicts in the error message file. Please resolve.

**Review by gkodinov — APPROVED** [review]
LGTM. Please resolve the merge conflicts and wait for the final review.

File: mysql-test/main/func_math.result [hunk 14]
```diff
@@ -45,9 +45,9 @@ select log(exp(10)),exp(log(sqrt(10))*2),log(-1),log(NULL),log(1,1),log(3,9),log
 log(exp(10))	exp(log(sqrt(10))*2)	log(-1)	log(NULL)	log(1,1)	log(3,9)	log(-1,2)	log(NULL,2)
 10	10.000000000000002	NULL	NULL	NULL	2	NULL	NULL
 Warnings:
-Warning	1365	Division by 0
-Warning	1365	Division by 0
-Warning	1365	Division by 0
+Warning	4257	Passing non-positive argument to log() makes the function's result undefined
+Warning	4258	Passing non-positive or equal-to-1 base to log() makes the function's result undefined
+Warning	4258	Passing non-positive or equal-to-1 base to log() makes the function's result undefined
```
Comment (vuvova) [inline mysql-test/main/func_math.result]: Let's not create too many very specific errors. I've just found `ER_INVALID_ARGUMENT_FOR_LOGARITHM` — let's use it for all these cases.

**Review by vuvova — CHANGES_REQUESTED** [review]

File: mysql-test/main/func_math.result (hunk 14 above)
Comment (alexaustin007) [reply mysql-test/main/func_math.result]: Addressed

File: sql/item_func.cc [hunk 15]
```diff
@@ -2021,7 +2030,8 @@ double Item_func_ln::val_real()
     return 0.0;
   if (value <= 0.0)
   {
-    signal_divide_by_null();
+    signal_log_domain_error(ER_INVALID_ARGUMENT_FOR_LOGARITHM);
```
Comment (vuvova) [inline sql/item_func.cc:2033]: eh, sorry for confusion. This patently stupid error message came from MySQL where it should've never been added in the first place. We shouldn't use it, sorry for suggesting it earlier, my mistake.

Please use `ER_STD_INVALID_ARGUMENT` which we consistently use everywhere. And `ErrConvDouble` for `value`

**Review by vuvova — CHANGES_REQUESTED** [review]

File: sql/item_func.cc (hunk 15 above)
Comment (alexaustin007) [reply sql/item_func.cc]: Addressed. Requesting to review

**Review by vuvova — APPROVED** [review]
Looks good, please squash into one commit.

I'd also suggest to make `signal_log_domain_error()` non-static method, then you can set `null_value=1` directly there, not in every caller. Or even you can define `signal_log_domain_error()` to return double 0.0, then it can be used as
```c
if (value <= 0.0)
  return signal_log_domain_error(value, func_name());
```
4 lines less to read at every call site.

Anyway, it's just a suggestion. Approved either way.

**Comment by alexaustin007** [comment]
> Looks good, please squash into one commit.
> 
> I'd also suggest to make `signal_log_domain_error()` non-static method, then you can set `null_value=1` directly there, not in every caller. Or even you can define `signal_log_domain_error()` to return double 0.0, then it can be used as
> 
> ```c
> if (value <= 0.0)
>   return signal_log_domain_error(value, func_name());
> ```
> 
> 4 lines less to read at every call site.
> 
> Anyway, it's just a suggestion. Approved either way.

Thank You!, Also dropped the `func_name()` argument since the helper is now a method on `Item_func` and already knows it. Requesting final review.

**Review by vuvova — APPROVED** [review]
